core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java

```diff
@@ -0,0 +1,288 @@
+      // 2. Master sends a registration command during
+      // the heartbeat and resets the registration state.
+      LOG.info("The worker needs to register with master {}", mMasterAddress);
+      // This will retry indefinitely and essentially block here if the master is not ready
+      registerWithMaster();
+      LOG.info("BlockMasterSync to master {} has started", mMasterAddress);
+    }
+    if (mWorkerState == WorkerMasterRegistrationState.REGISTERING) {
+      return;
+    }
+
+    RetryPolicy endlessRetry = createEndlessRetry();
+    while (endlessRetry.attempt()) {
+      BlockHeartbeatReport report = mBlockHeartbeatReporter.generateReportAndClear();
+      boolean success = false;
+      try {
+        beforeHeartbeat();
+        success = mBlockMasterSyncHelper.heartbeat(
+            mWorkerId.get(), report,
+            mBlockWorker.getStoreMeta(), this::handleMasterCommand);
+      } catch (Exception e) {
+        LOG.error("Failed to receive master heartbeat command. worker id {}", mWorkerId, e);
+      }
+      if (success) {
+        mLastSuccessfulHeartbeatMs = CommonUtils.getCurrentMs();
+        break;
+      } else {
+        LOG.warn(
+            "Heartbeat failed, worker id {}, worker host {} # of attempts {}, last success ts {}",
+            mWorkerId.get(), mWorkerAddress.getHost(), endlessRetry.getAttemptCount(),
+            mLastSuccessfulHeartbeatMs);
+        if (report.getBlockCount() >= mWorkerBlockHeartbeatReportCapacityThreshold) {
+          // If the report becomes too big, merging it back to the reporter might cause OOM issue.
+          // We throw away the result and let the worker re-register with the master.
+          mWorkerState = WorkerMasterRegistrationState.NOT_REGISTERED;
+          return;
+        } else {
+          mBlockHeartbeatReporter.mergeBack(report);
+        }
+      }
+    }
+  }
+
+  protected void beforeHeartbeat() {
+  }
+
+  @Override
+  public void close() {
+    mAsyncBlockRemover.shutDown();
+    mMasterClient.close();
+  }
+
+  /**
+   * @return if the worker has registered with the master successfully
+   */
+  public boolean isRegistered() {
+    return mWorkerState == WorkerMasterRegistrationState.REGISTERED;
+  }
+
+  /**
+   * Handles a master command. The command is one of Unknown, Nothing, Register, Free, or Delete.
+   * This call will block until the command is complete.
+   *
+   * @param cmd the command to execute
+   * @throws IOException               if I/O errors occur
+   * @throws ConnectionFailedException if connection fails
+   */
+  private void handleMasterCommand(Command cmd) throws IOException, ConnectionFailedException {
+    if (cmd == null) {
+      return;
+    }
+    switch (cmd.getCommandType()) {
+      // Currently unused
+      case Delete:
+        break;
+      // Master requests blocks to be removed from Alluxio managed space.
+      case Free:
+        mAsyncBlockRemover.addBlocksToDelete(cmd.getDataList());
```

Comment:
interestingly, double check what we do if the master is standby?

Reply:
yeah this is a good catch. i updated DefaultBlockMaster and made the standby master never return a Free command